### PR TITLE
[all] Enable govulncheck linter for the whole repo

### LINKED ISSUE #19154: [all] Enable govulncheck linter
### Component(s)

_No response_

### Describe the issue you're reporting

As a part of work in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9109 it would be beneficial to have [govulncheck](https://go.dev/blog/vuln) scans running in CI as well.

I think it's a great tool since it's part of the official Go toolset and as they state:
> Go provides tooling to analyze your codebase and surface known vulnerabilities. This tooling is backed by the Go vulnerability database, which is curated by the Go security team. Go’s tooling reduces noise in your results by only surfacing vulnerabilities in functions that your code is actually calling.

Also, please note it's NOT a part of `golangci-lint` (yet): https://github.com/golangci/golangci-lint/issues/3094.

### PR DESCRIPTION
**Pull request by lootek**
**Description:** <Describe what has changed.>

Enable [govulncheck](https://go.dev/blog/vuln) linter

**Link to tracking Issue:** <Issue number if applicable>
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19154

**Testing:** <Describe what testing was performed and which tests were added.>
N/A

**Documentation:** <Describe the documentation added.>
N/A

### CODE REVIEW

**Comment by lootek** [comment]
Looks like the GH Actions workflows won't run until I actually issue this PR as a non-draft one... So marking this as ready for review, yet I still need to see if it works as intended. I'll leave here a note when it's finally ready to be reviewed/merged.

**Review by atoulme — APPROVED** [review]

**Review by djaglowski — APPROVED** [review]

**Comment by lootek** [comment]
Despite the `go.sum` file changes, GH actions take Go tools stuff from cache and I keep getting
```
/bin/bash: line 1: /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/.tools/govulncheck: No such file or directory
```
https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4409057715/jobs/7726357708#step:9:18

Have you been hit by a similar issue in the past?

File: Makefile.Common [hunk 1]
```diff
@@ -162,9 +163,13 @@ fmt: $(GOIMPORTS)
 	$(GOIMPORTS) -w  -local github.com/open-telemetry/opentelemetry-collector-contrib ./
 
 .PHONY: lint
-lint: $(LINT) checklicense misspell
+lint: $(LINT) checklicense misspell govulncheck
 	$(LINT) run --allow-parallel-runners --build-tags integration
 
+.PHONY: govulncheck
+govulncheck:
```
Comment (codeboten) [inline Makefile.Common:170]:
```suggestion
govulncheck: $(GOVULNCHECK)
```

**Comment by codeboten** [comment]
> Have you been hit by a similar issue in the past?

The only time i seem to recall running into this was when `install-tools` hadn't run or was maybe using a cached version of the github step

**Comment by lootek** [comment]
> > Have you been hit by a similar issue in the past?
> 
> The only time i seem to recall running into this was when `install-tools` hadn't run or was maybe using a cached version of the github step

Yeah, so the very same issue I was hit with here... Do you happen to remember what was the solution? Or how can I force tools installation / cache refresh?

File: Makefile.Common (hunk 1 above)
Comment (lootek) [reply Makefile.Common:170]: Thank you :) But why do you suggest that? My reasoning was this is a command invoke so I placed that in its own indented line below, not as a dependency target in the same line.

As a side note, I bet `$(GOVULNCHECK) ./...` is the correct call.

Comment (codeboten) [reply Makefile.Common:170]: Adding that dependency will ensure the `govulncheck` binary is installed before running the make target `govulncheck`

Comment (lootek) [reply Makefile.Common:170]: silly me, you're right, thank you so much for help! :)

**Review by codeboten — APPROVED** [review]

**Comment by codeboten** [comment]
looks like govulncheck is finding issues, rather than fixing them all here and ending with a giant pr, i'd recommend opening separate issues for each one and adding an exception for each issue found?

```
Error:       alerts.go:235:3: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver.alertsReceiver.startListening calls github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver.startListening, which eventually calls crypto/internal/nistec.P256Point.ScalarBaseMult
Error:       alerts.go:253:3: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver.alertsReceiver.startListening calls github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver.startListening, which eventually calls crypto/internal/nistec.P256OrdInverse
Error:       alerts.go:253:3: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver.alertsReceiver.startListening calls github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver.startListening, which eventually calls crypto/internal/nistec.P256Point.ScalarMult
```

**Comment by lootek** [comment]
> looks like govulncheck is finding issues, rather than fixing them all here and ending with a giant pr, i'd recommend opening separate issues for each one and adding an exception for each issue found?
> 
> ```
> Error:       alerts.go:235:3: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver.alertsReceiver.startListening calls github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver.startListening, which eventually calls crypto/internal/nistec.P256Point.ScalarBaseMult
> Error:       alerts.go:253:3: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver.alertsReceiver.startListening calls github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver.startListening, which eventually calls crypto/internal/nistec.P256OrdInverse
> Error:       alerts.go:253:3: github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver.alertsReceiver.startListening calls github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbatlasreceiver.startListening, which eventually calls crypto/internal/nistec.P256Point.ScalarMult
> ```

Yeah, for sure, it'd be much more readable this way if that required some code changes. Yet the `vulncheck` is all about cross-checking the used dependencies' versions (including the standard Go library) against the known CVE-s database and it seems by the time I managed to have this scanner in the pipeline there were some findings and patch releases to Go. Please note it says there a few lines above:

```
Your code is affected by 1 vulnerability from the Go standard library.

Vulnerability #1: GO-2023-1621
  The ScalarMult and ScalarBaseMult methods of the P256 Curve may
  return an incorrect result if called with some specific
  unreduced scalars (a scalar larger than the order of the curve).
  This does not impact usages of crypto/ecdsa or crypto/ecdh.

 More info: https://pkg.go.dev/vuln/GO-2023-1621

 Standard library
    Found in: crypto/internal/nistec@go1.19.6
    Fixed in: crypto/internal/nistec@go1.20.2
```

So the actual "fix" would be to bump the Go version we use here from `1.19.6` to `1.20.2`. Do we have a special procedure for such action or can I do this here in my PR?

**Comment by lootek** [comment]
I gave it a try here, let's see if everything passes (as it should).

I've found https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18762 where the Go version was bumped recently so I guess it's just about changing those values and issuing a PR? :)

**Comment by Aneurysm9** [comment]
> So the actual "fix" would be to bump the Go version we use here from `1.19.6` to `1.20.2`. Do we have a special procedure for such action or can I do this here in my PR?

I think `1.19.7` would be the next version to use to resolve this.  Moving to 1.20 should be done separately and we've got workflows that need to use 1.19 anyways to ensure compatibility.  See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/18762 for an example of updating the workflows to use `1.19.6`.  I'd suggest doing it in this PR to avoid needing to introduce and later remove exceptions.

**Comment by lootek** [comment]
> I think `1.19.7` would be the next version to use to resolve this. Moving to 1.20 should be done separately and we've got workflows that need to use 1.19 anyways to ensure compatibility. See #18762 for an example of updating the workflows to use `1.19.6`. I'd suggest doing it in this PR to avoid needing to introduce and later remove exceptions.

Thanks @Aneurysm9 , this perfectly makes sense! I just made changes according to your suggestion, let's see if `govulncheck` passes this time :)

**Review by Aneurysm9 — APPROVED** [review]

**Comment by lootek** [comment]
The `vulncheck` lint scan step is passing now. Thank you everyone for approving, but I don't have the authorization to merge. Could you please do it for me when the time is right?

**Review by codeboten — APPROVED** [review]

**Pull request merged**